Blog: Adventures in null handling

ivanagas opened the pull request:
## Changes

A technical blog about Marius and Michael's adventures in null handling. Would appreciate feedback on any areas that are unclear or how to make more interesting.

## Checklist
- [x] Titles are in [sentence case](https://apastyle.apa.org/style-grammar-guidelines/capitalization/sentence-case)
- [x] Feature names are in **[sentence case too]([https://apastyle.apa.org/style-grammar-guidelines/capitalization/title-case](https://apastyle.apa.org/style-grammar-guidelines/capitalization/sentence-case))**
- [x] Words are spelled using American English
- [x] I have checked out our [style guide](https://github.com/PostHog/posthog.com/blob/master/STYLEGUIDE.md)

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,82 @@
+
+The value (or non-value) null can be the bane of a programmer’s existence. Null can work as expected in one language and context, then act in completely unexpected ways in another language or context.
+
+We recently had challenges with null during the [release of HogQL](/blog/introducing-hogql). Our engineers Marius and Michael realized null values for properties were leading to inaccurate results and customer confusion. This post covers their efforts to figure out and implement a solution for this.
+
+## The problem with null in HogQL
+
+The issue with null in HogQL was that SQL is strict about nulls. If any value in an expression is null, then the entire expression is null. Here are two HogQL expressions showcasing why this is a problem:
+
+1. `sumIf(properties.money, event = 'no problems') + sumIf(properties.money - 10, event = 'some problems')` returns null if either of the sides of the equation are null. This  happens if there are no matching events (named `no problems` or `some problems`) or the property (`money`) is null.
+2. `concat('money: ', properties.money)` returns null if there's no `money` property, though users might expect either `'money: '`  or `'money: null'`
+
+PostHog’s existing solutions for handling null did not work with the full access HogQL provides. Specifically:
+
+- ClickHouse skips null during aggregation, but users could want to concat or check null values with HogQL.
+- PostHog previously treated nulls as strings, but this could include null values in counts when they shouldn’t be.
+
+At this stage in HogQL’s development, if a value could be null, users must handle it themselves. For example, users must wrap each `sumIf` function in the first example above in a `nullIf(x, 0)` like this:
```
I feel there is some backgruond context missing. Specifically:

- What is HogQL?
- Tech stack e.g., clickhouse etc. 

It was a bit confusing to jump straight into things

ivanagas replied:
Added a half sentence on both of these.

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,82 @@
+
+## Brainstorming potential solutions
+
+To solve the issue with nulls, Marius and Michael came up with a bunch of potential solutions, (none of which were inevitably chosen):
+
+1. **Returning a default value depending on type.** For example, returning `""` for strings, `0` for numbers, or `false` for booleans. This didn’t handle the expression `properties.bla is null` and required rewriting `is set` and `is not set` logic.
+
+2. **Including a nullable checkbox in data management.** Similar to the previous option, but required users to input if they expect a property to be null. This required a lot of user work and remains confusing if they didn’t understand how to handle null.
+
+3. **Create JS-like `'bla' (NOT) IN properties` expression logic.** We felt this was too complex for users and makes the question of `is null` for an unset property unclear. 
+
+4. **Use the ClickHouse function `sumOrDefaultIf`.** Faces the same problem of being null if any values are null.
+
+5. **Use non-ANSI compliant string operator.** ANSI is SQL standards for function and syntax. For example, `f'{NULL}-haha-{properties.bla}'` is non-ANSI compliant because it doesn’t conform to standard SQL string concatenation syntax. We decided it is better to keep the standard operators and be flexible with types instead.
```
```suggestion
To solve the issue with nulls, Marius and Michael came up with a few potential solutions – but each solution had its own pitfalls!

1. **Returning a default value depending on type.** For example, returning `""` for strings, `0` for numbers, or `false` for booleans. This didn’t handle the expression `properties.bla is null` and required rewriting `is set` and `is not set` logic.

2. **Including a nullable checkbox in data management.** Similar to the previous option, but required users to input if they expect a property to be null. This required a lot of user work and remains confusing if they didn’t understand how to handle null.

3. **Create JS-like `'bla' (NOT) IN properties` expression logic.** We felt this was too complex for users and makes the question of `is null` for an unset property unclear. 

4. **Use the ClickHouse function `sumOrDefaultIf`.** Faces the same problem of being null if any values are null.

5. **Use non-ANSI compliant string operator.** ANSI is SQL standards for function and syntax. For example, `f'{NULL}-haha-{properties.bla}'` is non-ANSI compliant because it doesn’t conform to standard SQL string concatenation syntax. We decided it is better to keep the standard operators and be flexible with types instead.
```

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,82 @@
+
+## The solution(s) to our null problems
+
+The solution was to handle null for values we expect to be strings differently than numbers.
```
```suggestion
The solution was to handle null for values we expect to be strings differently than those that are expected to be numbers.
```

mariusandra replied:
This solution is very broad. Our solution was very simple: modify how `concat` works (it's also used if you add strings together with the `||` operator).

Previously pre-hogql everything was an empty string if null, but this is not correct. Ever since we introduced nulls, the biggest painpoint has been with concat-ing values. 

This `concat` function accepts only strings. However users want to join together strings, numbers, nulls and other values. This was the biggest pain point with nulls. So we made concat also accept numbers, nulls, dates and anything else. We didn't change any other function.

ivanagas replied:
reworded to be more specific about concat.

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,82 @@
+
+By understanding the common use cases and objectives of users when interacting with nulls, Marius implemented these solutions to streamline the user experience. He also didn’t implement a solution moving us too far away from standard practice, making HogQL more maintainable for us and understandable for users.
+
+## Further reading
```
Feels like a bit of an abrupt ending. From the title, I was expecting more scenarios and edge cases. Perhaps from the term "adventures", I was expecting more.

Can we add more to this post? Or maybe rename the title so that it sets expectations appropriately?

e.g. "Why null handling in SQL and Clickhouse is super annoying"

mariusandra commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,82 @@
+
+## The problem with null in HogQL
+
+The issue with null in HogQL was that SQL is strict about nulls. If any value in an expression is null, then the entire expression is null. Here are two HogQL expressions showcasing why this is a problem:
```
```suggestion
The issue with null in HogQL was that ClickHouse is strict about nulls. If any value in an expression is null, then [the entire expression is null](https://clickhouse.com/docs/en/sql-reference/functions#null-processing). Here are two HogQL expressions showcasing why this is a problem:
```

mariusandra commented on contents/blog/null-handling-hogql.md:
````diff
@@ -0,0 +1,82 @@
+At this stage in HogQL’s development, if a value could be null, users must handle it themselves. For example, users must wrap each `sumIf` function in the first example above in a `nullIf(x, 0)` like this:
+
+```
+nullIf(
+	sumIf(properties.money, event = 'no problems'), 
+	0
+) + nullIf(
+	sumIf(properties.money - 10, event = 'some problems'),
+	0
+)
````
these should actually be `ifNull`

mariusandra commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,82 @@
+
+To solve the issue with nulls, Marius and Michael came up with a bunch of potential solutions, (none of which were inevitably chosen):
+
+1. **Returning a default value depending on type.** For example, returning `""` for strings, `0` for numbers, or `false` for booleans. This didn’t handle the expression `properties.bla is null` and required rewriting `is set` and `is not set` logic.
```
```suggestion
1. **Returning a default value depending on type.** For example, returning `""` for strings, `0` for numbers, or `false` for booleans. This didn’t handle the expression `properties.bla is null` and required rewriting `is set` and `is not set` logic. Plus default zeroes might mess up aggregations, when you expect null fields to be discarded.
```

(or something to that extend)

mariusandra commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,82 @@
+
+3. **Create JS-like `'bla' (NOT) IN properties` expression logic.** We felt this was too complex for users and makes the question of `is null` for an unset property unclear. 
+
+4. **Use the ClickHouse function `sumOrDefaultIf`.** Faces the same problem of being null if any values are null.
```
This is not exactly true. We will likely expose all the aggregations variations like `sum` (exposed), `sumIf` (exposed), `sumOrDefault`, `sumOrDefaultIf`, and perhaps some more. TODO

Also, it'll be null if _all_ values are nulls 🤷 

```sql
select sumOrDefault(a) from (select arrayJoin([1,2,3, null]) as a) -- returns 6
select sumOrDefault(a) from (select arrayJoin([null, null, null]) as a) -- returns null, NOT 1
```

Also,

```sql
select avg(a) from (select arrayJoin([1,2,3, null]) as a) -- returns 2 as nulls are just ignored
```

mariusandra commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,82 @@
+
+Strings don’t need the same accuracy, users mostly want to return or concat strings anyway. To solve the specific issue with the concat function, Marius made it null-tolerant. This means every value converts to a string. For example, `concat(null, 'a', 3, toString(4), toString(NULL), properties.$screen_width)` turns into `concat('', 'a', toString(3), toString(4), '')`. 
+
+To do this, PostHog automatically adds `ifNull` falling back to empty strings (`''`) to concat functions. HogQL abstracts this away because the functionality creates a long and "dorky" ClickHouse SQL strings like:
```
> "a long and dorky strings"

singular "a" and plural "-s" don't match

mariusandra reviewed (commented):
Added some technical context

mariusandra commented:
Possible pivot 🔄 - convert this to a bigger post on "the tweaks we did to clickhouse", include
- the bigger point: we can enhance clickhouse functions, add our own, or even introduce new language constructs if that makes things easier
- enhancement: (this blog post), making string concatenation with `concat()` accept nulls and numbers, because, that just makes sense for our case
- addition: https://github.com/PostHog/posthog/pull/16096
- addition and a new language construct (`in cohort 'my cohort'`): https://github.com/PostHog/posthog/pull/16119

ivanagas commented:
Addressed all the points. I think I will keep this focused on null handling, because there isn't enough on other modifications to HogQL (but it is now the conclusion).

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,88 @@
+
+The value (or non-value) null can be the bane of a programmer’s existence. Null can work as expected in one language and context, then act in completely unexpected ways in another language or context.
+
+We had challenges with null during the [release of HogQL](/blog/introducing-hogql), which added direct SQL access to your data. Our engineers Marius and Michael realized null values for properties were leading to inaccurate results and customer confusion. This post covers their efforts to figure out and implement a solution for this.
```
```suggestion
We had challenges working with null during the [release of HogQL](/blog/introducing-hogql), which added direct SQL access to your data. Our engineers Marius and Michael realized null values for properties were leading to inaccurate results and customer confusion. This post covers their efforts to figure out and implement a solution for this.
```

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,88 @@
+
+## The problem with null in HogQL
+
+The issue with null in HogQL was that [ClickHouse](/docs/how-posthog-works/clickhouse), the database we use to store and process data, is strict about nulls. If any value in an expression is null, then [the entire expression is null](https://clickhouse.com/docs/en/sql-reference/functions#null-processing). Here are two HogQL expressions showcasing why this is a problem:
```
```suggestion
The issue with null in HogQL was that [ClickHouse](/docs/how-posthog-works/clickhouse), the database we use to store and process data, is **strict** about nulls. If any value in an expression is null, then [the entire expression is null](https://clickhouse.com/docs/en/sql-reference/functions#null-processing). Here are two HogQL expressions showcasing why this is a problem:
```

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,88 @@
+
+The issue with null in HogQL was that [ClickHouse](/docs/how-posthog-works/clickhouse), the database we use to store and process data, is strict about nulls. If any value in an expression is null, then [the entire expression is null](https://clickhouse.com/docs/en/sql-reference/functions#null-processing). Here are two HogQL expressions showcasing why this is a problem:
+
+1. `sumIf(properties.money, event = 'no problems') + sumIf(properties.money - 10, event = 'some problems')` returns null if either of the sides of the equation is null. This  happens if there are no matching events (named `no problems` or `some problems`) or the property (`money`) is null.
```
```suggestion
1. `sumIf(properties.money, event = 'no problems') + sumIf(properties.money - 10, event = 'some problems')` returns null if either of the sides of the equation is null. This happens if there are no matching events (named `no problems` or `some problems`) or the property (`money`) is null.
```

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,88 @@
+The issue with null in HogQL was that [ClickHouse](/docs/how-posthog-works/clickhouse), the database we use to store and process data, is strict about nulls. If any value in an expression is null, then [the entire expression is null](https://clickhouse.com/docs/en/sql-reference/functions#null-processing). Here are two HogQL expressions showcasing why this is a problem:
+
+1. `sumIf(properties.money, event = 'no problems') + sumIf(properties.money - 10, event = 'some problems')` returns null if either of the sides of the equation is null. This  happens if there are no matching events (named `no problems` or `some problems`) or the property (`money`) is null.
+2. `concat('money: ', properties.money)` returns null if there's no `money` property, though users might expect either `'money: '`  or `'money: null'`
```
```suggestion
2. `concat('money: ', properties.money)` returns null if the `money` property does not exist, though users might expect either `'money: '`  or `'money: null'`
```

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,88 @@
+---
+title: "Adventures in null handling: Null money, null problems"
```
lol

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,88 @@
+
+## Brainstorming potential solutions
+
+To solve the issue with nulls, Marius and Michael came up with a few potential solutions, but each had pitfalls.
```
```suggestion
To solve the issue with nulls, Marius and Michael came up with a few potential solutions, but each had pitfalls:
```

Lior539 commented on contents/blog/null-handling-hogql.md:
```diff
@@ -0,0 +1,88 @@
+
+Handling numbers remained the same because returning `0` for null number properties meaningfully affects calculations; they get added as non-null to counts. To prevent this, PostHog continues to return null and lets ClickHouse skip these in aggregations.
+
+## Modifying SQL with HogQL
```
Maybe change this to ## Conclusion ?

Lior539 reviewed (approved):
Much better! I enjoyed reading this!
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia3.giphy.com%2Fmedia%2FpCO5tKdP22RC8%2Fgiphy.gif&token=nice" data-gitmeme-token="nice"><img src="https://media3.giphy.com/media/pCO5tKdP22RC8/giphy.gif" title="Created by gitme.me with /nice"/></a>

andyvan-ph commented:
A few thoughts:

- I think this would be better as a ghost written piece – i.e. it's not "Marius and Michael did X", it's "we did this" or "I did this".

- We should very quickly explain what HogQL is in the intro. People shouldn't need prior knowledge, or have to click away, to understand the context.

- I think the final section / conclusion should really be about "why are we doing this?" or, in other words, why should the reader care about this?

ivanagas commented:
Now "ghost-written" and rewrote the conclusion again, ready for your review @andyvan-ph

mariusandra commented:
Might be relevant to add that we also:

1. added a [nullish coalescing](https://github.com/PostHog/posthog/pull/16276) option... So now you can do `properties.bla ?? properties.another_one ?? properties.moop` and the first non-null wins.
2. fixed [null equality operations](https://github.com/PostHog/posthog/pull/16259) that were causing problems
